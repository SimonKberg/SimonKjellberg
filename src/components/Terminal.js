// @flow strict
import * as React from 'react'
import screenfull from 'screenfull'
import styled, { css } from 'react-emotion'
import { theme, withProp } from 'styled-tools'
import { darken } from 'polished'

const Window = styled('div')`
  position: relative;
  display: flex;
  flex-direction: column;
  margin: 0 auto;
  width: 100%;
  margin-top: 1em;
  margin-bottom: 1em;
  border-radius: 0.625rem;
  background-color: ${theme('terminal.window.backgroundColor')};

  &:fullscreen {
    height: 100%;
    margin: 0;
  }

  &:-ms-fullscreen {
    height: 100%;
    margin: 0;
  }

  &:-moz-full-screen {
    height: 100%;
    margin: 0;
  }

  &:-webkit-full-screen {
    height: 100%;
    margin: 0;
  }
`

const WindowTopbar = styled('div')`
  background-color: ${theme('terminal.topbar.backgroundColor')};
  height: ${theme('terminal.topbar.height')};
  border-radius: 0.5rem 0.5rem 0 0;
  padding: 0 0.625rem;
`

const windowControl = css`
  display: inline-block;
  width: 0.75rem;
  height: 0.75rem;
  padding: 0;
  margin: 0.625rem 0.25rem 0 0;
  border-radius: 0.5rem;
  border: 1px solid transparent;

  &:focus {
    outline: none;
  }
`

const WindowControlClose = styled('button')`
  ${windowControl};
  background-color: ${theme('terminal.controls.close.backgroundColor')};

  &:focus {
    border-color: ${withProp(
      'theme.terminal.controls.close.backgroundColor',
      darken(0.15)
    )};
  }
`

const WindowControlMinimize = styled('button')`
  ${windowControl};
  background-color: ${theme('terminal.controls.minimize.backgroundColor')};

  &:focus {
    border-color: ${withProp(
      'theme.terminal.controls.minimize.backgroundColor',
      darken(0.15)
    )};
  }
`

const WindowControlMaximize = styled('button')`
  ${windowControl};
  background-color: ${theme('terminal.controls.maximize.backgroundColor')};

  &:focus {
    border-color: ${withProp(
      'theme.terminal.controls.maximize.backgroundColor',
      darken(0.15)
    )};
  }
`

const WindowContent = styled('div')`
  display: flex;
  flex-direction: column;
  position: relative;
  color: ${theme('terminal.content.color')};
  max-height: ${theme('terminal.content.maxHeight')};
  margin: 0 1.25rem;

  ${Window}:fullscreen & {
    max-height: none;
  }

  ${Window}:-ms-fullscreen & {
    max-height: none;
  }

  ${Window}:-moz-full-screen & {
    max-height: none;
  }

  ${Window}:-webkit-full-screen & {
    max-height: none;
  }
`

type Props = {
  children?: React.Node,
}

const windowRef = React.createRef()

const onClickMaximize = () => {
  if (screenfull.enabled) {
    screenfull.toggle(windowRef.current)
  }
}

const Terminal = ({ children }: Props) => (
  <Window innerRef={windowRef}>
    <WindowTopbar>
      <WindowControlClose />
      <WindowControlMinimize />
      <WindowControlMaximize onClick={onClickMaximize} title="Maximize" />
    </WindowTopbar>
    <WindowContent>{children}</WindowContent>
  </Window>
)

export default Terminal
