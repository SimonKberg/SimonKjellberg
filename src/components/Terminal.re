module Styles = {
  open Css;

  let windowFullscreen = [height(pct(100.0)), margin(`zero)];

  let window =
    style([
      position(`relative),
      display(`flex),
      flexDirection(`column),
      margin2(~v=em(1.0), ~h=`auto),
      width(pct(100.0)),
      borderRadius(rem(0.625)),
      backgroundColor(Theme.Terminal.Window.backgroundColor),
      selector("&:fullscreen", windowFullscreen),
      selector("&:-ms-fullscreen", windowFullscreen),
      selector("&:-moz-full-screen", windowFullscreen),
      selector("&:-webkit-full-screen", windowFullscreen),
    ]);

  let windowTopbar =
    style([
      backgroundColor(Theme.Terminal.Topbar.backgroundColor),
      height(Theme.Terminal.Topbar.height),
      borderTopLeftRadius(rem(0.5)),
      borderTopRightRadius(rem(0.5)),
      padding2(~v=`zero, ~h=rem(0.625)),
    ]);

  let windowControl =
    style([
      display(`inlineBlock),
      width(rem(0.75)),
      height(rem(0.75)),
      padding(`zero),
      marginTop(rem(0.625)),
      marginRight(rem(0.25)),
      borderRadius(rem(0.5)),
      border(px(1), `solid, `transparent),
      focus([outlineStyle(`none)]),
    ]);

  let windowControlClose =
    merge([
      windowControl,
      style([
        backgroundColor(Theme.Terminal.Controls.Close.backgroundColor),
        focus([
          borderColor(
            Polished.darken(
              0.15,
              Theme.Terminal.Controls.Close.backgroundColor,
            ),
          ),
        ]),
      ]),
    ]);

  let windowControlMinimize =
    merge([
      windowControl,
      style([
        backgroundColor(Theme.Terminal.Controls.Minimize.backgroundColor),
        focus([
          borderColor(
            Polished.darken(
              0.15,
              Theme.Terminal.Controls.Minimize.backgroundColor,
            ),
          ),
        ]),
      ]),
    ]);
  let windowControlMaximize =
    merge([
      windowControl,
      style([
        backgroundColor(Theme.Terminal.Controls.Maximize.backgroundColor),
        focus([
          borderColor(
            Polished.darken(
              0.15,
              Theme.Terminal.Controls.Maximize.backgroundColor,
            ),
          ),
        ]),
      ]),
    ]);

  let windowContentFullscreen = [
    maxHeight(Calc.(-)(vh(100.), Theme.Terminal.Topbar.height)),
  ];

  let windowContent =
    style([
      display(`flex),
      flexDirection(`column),
      position(`relative),
      color(Theme.Terminal.Content.color),
      maxHeight(Theme.Terminal.Content.maxHeight),
      margin2(~v=`zero, ~h=rem(0.625)),
      selector("." ++ window ++ ":fullscreen &", windowContentFullscreen),
      selector("." ++ window ++ ":-ms-fullscreen &", windowContentFullscreen),
      selector(
        "." ++ window ++ ":-moz-full-screen &",
        windowContentFullscreen,
      ),
      selector(
        "." ++ window ++ ":-webkit-full-screen &",
        windowContentFullscreen,
      ),
    ]);
};

[@react.component]
let make = (~children) => {
  let windowRef = React.useRef(Js.Nullable.null);
  let handleClickMaximize =
    React.useCallback1(
      _ =>
        if (Screenfull.isEnabled) {
          switch (windowRef->React.Ref.current->Js.Nullable.toOption) {
          | Some(r) => Screenfull.toggle(r) |> ignore
          | None => ()
          };
        },
      [|windowRef|],
    );

  <div className=Styles.window ref={ReactDOMRe.Ref.domRef(windowRef)}>
    <div className=Styles.windowTopbar>
      <button className=Styles.windowControlClose />
      <button className=Styles.windowControlMinimize />
      <button
        className=Styles.windowControlMaximize
        onClick=handleClickMaximize
      />
    </div>
    <div className=Styles.windowContent> children </div>
  </div>;
};

let default = make;
