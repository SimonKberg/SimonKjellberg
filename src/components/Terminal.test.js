import * as React from 'react'
import renderer from 'react-test-renderer'
import Terminal from './Terminal.bs'

test('<Terminal />', () => {
  const tree = renderer.create(<Terminal>Hello world!</Terminal>).toJSON()
  expect(tree).toMatchSnapshot()
})
