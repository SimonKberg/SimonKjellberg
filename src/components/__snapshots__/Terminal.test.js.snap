// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Terminal /> 1`] = `
.emotion-5 {
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 1em auto;
  width: 100%;
  border-radius: 0.625rem;
  background-color: #0d1f2d;
}

.emotion-5:fullscreen {
  height: 100%;
  margin: 0px;
}

.emotion-5:-ms-fullscreen {
  height: 100%;
  margin: 0px;
}

.emotion-5:-moz-full-screen {
  height: 100%;
  margin: 0px;
}

.emotion-5:-webkit-full-screen {
  height: 100%;
  margin: 0px;
}

.emotion-3 {
  background-color: #e0e8f0;
  height: 1.875rem;
  border-top-left-radius: 0.5rem;
  border-top-right-radius: 0.5rem;
  padding: 0 0.625rem;
}

.emotion-0 {
  display: inline-block;
  width: 0.75rem;
  height: 0.75rem;
  padding: 0px;
  margin-top: 0.625rem;
  margin-right: 0.25rem;
  border-radius: 0.5rem;
  border: 1px solid transparent;
  background-color: #ff6057;
}

.emotion-0:focus {
  outline-style: none;
}

.emotion-0:focus {
  border-color: #ff180b;
}

.emotion-1 {
  display: inline-block;
  width: 0.75rem;
  height: 0.75rem;
  padding: 0px;
  margin-top: 0.625rem;
  margin-right: 0.25rem;
  border-radius: 0.5rem;
  border: 1px solid transparent;
  background-color: #ffbd2e;
}

.emotion-1:focus {
  outline-style: none;
}

.emotion-1:focus {
  border-color: #e19a00;
}

.emotion-2 {
  display: inline-block;
  width: 0.75rem;
  height: 0.75rem;
  padding: 0px;
  margin-top: 0.625rem;
  margin-right: 0.25rem;
  border-radius: 0.5rem;
  border: 1px solid transparent;
  background-color: #27c93f;
}

.emotion-2:focus {
  outline-style: none;
}

.emotion-2:focus {
  border-color: #1b892b;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  color: #f4faff;
  max-height: 25rem;
  margin: 0 0.625rem;
}

.emotion-5:fullscreen .emotion-4 {
  max-height: calc(100vh - 1.875rem);
}

.emotion-5:-ms-fullscreen .emotion-4 {
  max-height: calc(100vh - 1.875rem);
}

.emotion-5:-moz-full-screen .emotion-4 {
  max-height: calc(100vh - 1.875rem);
}

.emotion-5:-webkit-full-screen .emotion-4 {
  max-height: calc(100vh - 1.875rem);
}

<div
  className="emotion-5"
>
  <div
    className="emotion-3"
  >
    <button
      className="emotion-0"
    />
    <button
      className="emotion-1"
    />
    <button
      className="emotion-2"
      onClick={[Function]}
    />
  </div>
  <div
    className="emotion-4"
  >
    Hello world!
  </div>
</div>
`;
